### ISSUE TITLE
YOLOv8 Classification Model from FP32 to FP16 and INT8

### ISSUE THREAD

**Issue by pornpra** [ultralytics/ultralytics#10397]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

After trained YOLOv8 Classification model (FP32), how can I convert it into FP16 and INT8 ?

**Comment by glenn-jocher** [comment]
Hello! 😊 

To convert your YOLOv8 Classification model from FP32 to FP16 and INT8, you can use the export functionality provided by the Ultralytics YOLO framework. Here's how you can perform the conversions:

For **FP16**:
```bash
yolo export model=path/to/your/model.pt format=onnx half=True
```

For **INT8** quantization, currently, exporting directly to INT8 is not supported directly via the Ultralytics export command for all models. However, once you have the ONNX model, you might consider using tools like the ONNX Runtime or TensorRT (for NVIDIA GPUs) which can further quantize the model to INT8 and provide inference capabilities.

Assuming you've already exported to ONNX as shown above, for NVIDIA GPUs, you can use TensorRT's quantization tooling to process the ONNX model into INT8, focusing on creating a calibration dataset and running the model optimizer with INT8 options. Please refer to the [TensorRT documentation](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html) for the specifics on this process.

Remember, moving to FP16 and especially INT8 may lead to a loss in accuracy but could provide significant improvements in inference speed and memory usage.

Hope this helps! If you have more questions regarding the process, we’re here to assist. 🚀

**Comment by MathijsNL** [comment]
Looking at the export documentation, int8 is supported by TF Lite and OpenVINO.

When trying to export with OpenVINO it is necessary to supply calibration data (not sure about the other methods).

A problem I ran into is that providing the dataset directory for classification task is not accepted by the export command:

```
yolo export model=best.pt format=openvino imgsz=320 int8=True batch=1 data=./calibration_images

Ultralytics YOLOv8.2.3 🚀 Python-3.11.8 torch-2.2.1+cu121 CPU (AMD Ryzen 7 7800X3D 8-Core Processor)
YOLOv8n-cls summary (fused): 73 layers, 1438723 parameters, 0 gradients, 3.3 GFLOPs

PyTorch: starting from 'best.pt' with input shape (1, 3, 320, 320) BCHW and output shape(s) (1, 3) (2.8 MB)

OpenVINO: starting export with openvino 2024.1.0-15008-f4afc983258-releases/2024/1...
INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, tensorflow, onnx, openvino
OpenVINO: collecting INT8 calibration images from 'data=./calibration_images/'
OpenVINO: export failure ❌ 0.7s: [Errno 21] Is a directory: './calibration_images/'
```

Is this because classification is not fully supported yet, or is there a way to supply calibration images for classification model? It seems that the model is actually working with OpenVINO format after exporting without calibration data without too much loss in accuracy.

**Comment by glenn-jocher** [comment]
Hey there! 👋

Great to hear you've been diving into the export options with YOLOv8, especially exploring INT8 quantization with OpenVINO! 🚀 When it comes to providing calibration data for the INT8 quantization process in OpenVINO, it indeed expects a specific format or path handling that may differ slightly from expected. 

For calibration data during an OpenVINO export, the expected input isn’t a direct folder path through the `data` argument but rather integrated as part of a more extensive configuration process. Unfortunately, directly specifying a dataset directory like this isn’t currently supported for the `yolo export` command for classification tasks in the manner you've tried.

As a workaround, after exporting your model to the OpenVINO format without INT8 calibration (as you've noted, with possibly minimal accuracy loss), you could manually run the model through the OpenVINO Model Optimizer with calibration data following the steps in their [calibration tools documentation](https://docs.openvino.ai/latest/workbench_docs_Workbench_DG_Int_8_Quantization.html).

It sounds like you've got a good handle on things even without the INT8 calibration dataset, which is awesome! Keep experimenting and stay tuned for any updates on our side that might streamline this process in the future.

If you discover any more insights or if there's anything else we can help with, feel free to reach out. Happy coding! 😊

**Comment by pornpra** [comment]
@glenn-jocher 
Thanks for your answers.

Could you share the preprocessing steps for YOLOv8 for classification, using PyTorch code? I need them to prepare for int8 calibration.

**Comment by glenn-jocher** [comment]
@pornpra hey there! 🌟

Absolutely, happy to clarify the preprocessing steps for YOLOv8 classification during inference. While `classify_transforms` is designed for augmentation, for inference preprocessing, you generally want to resize and normalize the input image. Below is a simplified example using PyTorch:

```python
from torchvision import transforms
from PIL import Image

# Load your image
image = Image.open("path/to/your/image.jpg")

# Define preprocessing transformation
preprocess = transforms.Compose([
    transforms.Resize((320, 320)),  # Resize to match model input
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406],  # Standard normalization
                         std=[0.229, 0.224, 0.225])
])

# Apply preprocessing
input_tensor = preprocess(image)
input_batch = input_tensor.unsqueeze(0)  # Create a mini-batch as expected by the model
```

This should align your image preparation with what's needed for INT8 calibration data. Modify the resize dimensions as per your model's requirements! 🛠️ 

Let me know if there's anything else I can help with. Happy coding!

**Comment by pornpra** [comment]
@glenn-jocher 

Thanks for the clarification. 

I have one more question regarding the standard normalization values. By default, are the mean and standard deviation set to (0.0, 0.0, 0.0) and (1.0, 1.0, 1.0) respectively for the YOLOv8 classification model during inference? If I am mistaken, could you share the correct values for mean and standard deviation?

**Comment by glenn-jocher** [comment]
@pornpra hey there! 😊

The default mean and standard deviation for normalization in the YOLOv8 classification model during inference are not set to zeros and ones. Instead, we typically use ImageNet standards for normalization, which are:

- Mean: `[0.485, 0.456, 0.406]`
- Standard Deviation: `[0.229, 0.224, 0.225]`

Here’s how you can apply this in your code:

```python
transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
```

This helps align your input data with the pre-processing steps the model expects, based on the training data distribution!

Hope this clears things up! Let me know if you have any more questions.

**Comment by pornpra** [comment]
Hi

I have already successfully converted a customized YOLOv8 classification model from FP32 to INT8. During inference, I've noticed a significant performance drop. I used 3,600 images as a calibration dataset with NVIDIA Python code for the conversion. Is this enough?

I used the same calibration set with other custom classification CNNs like ResNet, EfficientNet, and MobileNet, but their accuracy did not drop like YOLOv8’s

**Comment by MathijsNL** [comment]
How did you do the conversion? For my datasets the model performance in the test set was unchanged. I used OpenVINO as export format using int8.

**Comment by MathijsNL** [comment]
I think this Nvidia forum post might answer your question:
[https://forums.developer.nvidia.com/t/int8-calibration-in-python-with-tensorrt-8-6/257643/6](https://forums.developer.nvidia.com/t/int8-calibration-in-python-with-tensorrt-8-6/257643/6)

**Comment by glenn-jocher** [comment]
Hey there! 😊

Thanks for sharing this helpful NVIDIA forum post! It indeed provides valuable insights on the INT8 calibration process with TensorRT, which might be exactly what you're looking for. If you're diving into INT8 conversion for optimized inference, this resource should prove quite beneficial.

Here's a quick example of integrating a similar calibration approach with your YOLOv8 model for TensorRT, adapted from the discussion:

```python
# Assuming you've got your model loaded and ready for conversion

# Initialize INT8 calibration with the specified dataset
calibrator = MyCalibrator(data_dir="path/to/calibration_images")

# Convert the model
engine = builder.build_engine(network, config, calibrator)
```

Remember, the key to successful INT8 calibration is a representative calibration dataset. Feel free to tweak the dataset or calibration method as needed based on your specific use case.

Let me know if you need further assistance! 🚀

**Comment by glenn-jocher** [comment]
@MathijsNL @pornpra great news 😃! Your original issue may now be resolved ✅ in PR #10423 which is now publishing to `ultralytics 8.2.9`. 

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!
